### PR TITLE
feat: add `decorations` library with `brace` function

### PR DESCRIPTION
**Pull request by RubixDev**
This PR adds a new CeTZ library called "Decorations". Currently it only contains one function, i.e. `braces`, but I thought other decorations may also be added to the same module in the future. For example, zig-zag and coiled lines, or more complex shapes like gears.

Here is a screenshot of the example I added to the manual:

image

A few points you may want to consider before merging:

- The brace function does not define a custom style root, and because it uses `merge-path` internally also doesn't even use the `bezier` styles. I think it would be better for users if they could set styles for all braces, also including the things that are currently direct function parameters (`amplitude`, `pointiness`, `content-offset`). I just don't have enough experience with CeTZ yet to implement that myself.
- The `brace` function takes a `debug` parameter. Should this maybe just respect the canvas's `debug` setting (if that is even possible)? It shows more specific and different information than the normal `debug` though.
- The `brace` function currently does no type and value checking of the parameters. Should some assertions be added?
- The `brace` function exposes anchors for all debug points. I guess this could potentially expose a bit too much of the internal structure and may lead to breaking changes for just slight internal refactors. The most important points are already exposed with appropriate names, so I would have no problem with removing these anchors.
- I just saw that there is a `line-normal` function in the `util` module, which could be used in two places instead of the `_rotate-around` function and may be faster or more accurate.
- The current description of the module in the manual is a bit meh, but I don't really know what to write there with only one function existing yet.

Also just for reference, this PR started out with me experimenting on the Typst Discord [here](https://discord.com/channels/1054443721975922748/1056574658771681361/1157280088396668978).

### CODE REVIEW

**Review by fenjalien — CHANGES_REQUESTED** [review]
This is really impressive nice work!
- Have a look at the other libs to see how they resolve their styles
- If you wrap the whole thing in a group and use a callback you can get `ctx` which has `ctx.debug` so you could attach it to that. You can keep the parameter anyway as it is cool to see.
- Have added a comment regarding type checking, main values to focus are coordinates. You can do it without `ctx` by using `coordinate.resolve-system(...)` which will just return a string of the coordinate type or panics if it is invalid.
- Don't worry about the debug anchors they're cool.
- If you can use something that aready exists in then yea, if not dw.
- Its alright.

Could you add a test for them? we don't have any guidlines but @johannes-wolf would be able to help if you need it.

**Comment by Beiri22** [comment]
That is exactly what I need; please, when you merge this PR, push it as new version to the package repo!

**Comment by johannes-wolf** [comment]
> That is exactly what I need; please, when you merge this PR, push it as new version to the package repo!

Ok, willrelease 0.1.3 after this got merged.

**Comment by johannes-wolf** [comment]
This looks great! Cant wait to have this merged :)

File: src/lib/decorations.typ [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+///   / end:     End coordinate
+///   / spike:   Point of the spike
+///   / content: Point to place content at
+///   / center:  Center of the enclosing rectangle
+///   / (a-i):   Debug points
+///
+/// - a (coordinate): Start point
+/// - b (coordinate): End point
+/// - amplitude (int,float): Height of the brace
+/// - pointiness (angle): How pointy the spike should be. `0deg` for maximum pointiness, `90deg` for minimum.
```
[the remark this answers is not in the document]
Comment (RubixDev) [reply src/lib/decorations.typ:31]: yes `90deg` is a flat line. values outside that range give funky results, for example `180deg`:
image
I think I will allow the value to either be an angle between `0deg` and `90deg` or a number between 0 and 1

File: src/lib/decorations.typ [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+///   / start:   Start coordinate
+///   / end:     End coordinate
+///   / spike:   Point of the spike
+///   / content: Point to place content at
+///   / center:  Center of the enclosing rectangle
+///   / (a-i):   Debug points
+///
+/// - a (coordinate): Start point
+/// - b (coordinate): End point
+/// - amplitude (int,float): Height of the brace
```
[the remark this answers is not in the document]
Comment (RubixDev) [reply src/lib/decorations.typ:30]: are you sure I should use `number`? that is not an official Typst type. It should actually be `(integer, float)`

[the remark this answers is not in the document]
Comment (fenjalien) [reply src/lib/decorations.typ:30]: yes we use it elsewhere in the manual. also forgot to add: note that the default for this parameter in tikz is `2.5pt`. would this value work as a default here? you can resolve lengths by using `util.resolve-number`

File: src/lib/decorations.typ [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+/// *Anchors:*
+///   / start:   Start coordinate
+///   / end:     End coordinate
+///   / spike:   Point of the spike
+///   / content: Point to place content at
+///   / center:  Center of the enclosing rectangle
+///   / (a-i):   Debug points
+///
+/// - a (coordinate): Start point
+/// - b (coordinate): End point
```
[the remark this answers is not in the document]
Comment (RubixDev) [reply src/lib/decorations.typ:29]: done

File: src/lib/decorations.typ [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  return vector.add(a, scaled)
+}
+
+/// Draw a curly brace between two points.
+///
+/// *Anchors:*
+///   / start:   Start coordinate
+///   / end:     End coordinate
+///   / spike:   Point of the spike
+///   / content: Point to place content at
```
[the remark this answers is not in the document]
Comment (RubixDev) [reply src/lib/decorations.typ:24]: new version:
```typ
/// *Anchors:*
///   / start:   Where the brace starts, same as the `start` parameter.
///   / end:     Where the brace end, same as the `end` parameter.
///   / spike:   Point of the spike, halfway between `start` and `end` and shifted
///     by `amplitude` towards the pointing direction.
///   / content: Point to place content/text at, in front of the spike.
///   / center:  Center of the enclosing rectangle.
///   / (a-i):   Debug points `a` through `i`.
```

File: src/lib/decorations.typ (hunk 2 above)
[the remark this answers is not in the document]
Comment (RubixDev) [reply src/lib/decorations.typ:30]: and I adopted the description from TikZ

File: src/lib/decorations.typ [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+  b,
+  amplitude: .7,
+  pointiness: 15deg,
+  content-offset: .3,
+  flip: false,
+  debug: false,
+  name: none,
+  ..style,
+) = {
+  // TODO: custom style root
```
[the remark this answers is not in the document]
Comment (RubixDev) [reply src/lib/decorations.typ:48]: it works!

File: src/lib/decorations.typ [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  // TODO: type and value assertions
```
[the remark this answers is not in the document]
Comment (RubixDev) [reply src/lib/decorations.typ:49]: I added type and value checks for the three parameters `amplitude`, `pointiness`, and `content-offset` which have also all been moved to be styles

File: src/lib/decorations.typ [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  // flipping is achieved by swapping the start and end points, the parameter is just for convenience
```
[the remark this answers is not in the document]
Comment (RubixDev) [reply src/lib/decorations.typ:50]: yes, that's much simpler, thanks

File: src/lib/decorations.typ [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+      anchor(name, point)
+    }
+  }, name: name)
+
+  // label all points in debug mode
+  if debug {
+    for (name, point) in points {
+      content(point, box(fill: luma(240), inset: .5pt, text(6pt, raw(name))))
+    }
+  }
```
[the remark this answers is not in the document]
Comment (RubixDev) [reply src/lib/decorations.typ:123]: that's smart. done!

**Review by johannes-wolf — APPROVED** [review]
LGTM.
~~I will add tests for this in a follow-up PR.~~

**Comment by johannes-wolf** [comment]
@Beiri22 Release PR: https://github.com/typst/packages/pull/165

**Pull request merged**